File: easyamp/sources/plex.py
"""Plex source: PIN/link auth, server discovery, library browse, direct-play.

Auth is Plex's sanctioned third-party flow: create a PIN at plex.tv, send
the user to the link page to approve it, poll the PIN for an account
token, then exchange that for per-server access via the resources API.
Every request sends ``Accept: application/json`` (Plex answers XML
otherwise) and the standard ``X-Plex-*`` client headers.

Browse item ids encode the hierarchy as plain strings:
``sections`` → ``sec/<key>`` → ``artist/<ratingKey>`` →
``album/<ratingKey>``, plus ``playlists`` → ``playlist/<ratingKey>``.
"""

from __future__ import annotations

import socket
import urllib.error
import urllib.parse
import urllib.request

from .. import __version__
from .. import secrets_store
from .base import BrowseItem, MusicSource, SourceError, Track, http_json

PLEX_TV = "https://plex.tv"
_PRODUCT = "EasyAmp"


def _headers(client_id: str, token: str = "") -> dict:
    h = {"X-Plex-Product": _PRODUCT,
         "X-Plex-Version": __version__,
         "X-Plex-Client-Identifier": client_id,
         "X-Plex-Device-Name": _PRODUCT,
         "X-Plex-Platform": "GTK"}
    if token:
        h["X-Plex-Token"] = token
    return h


# ---- PIN / link auth (module-level: runs before an account exists) -------
def start_pin(client_id: str) -> tuple[int, str]:
    """Create a link PIN; returns (pin_id, 4-char code)."""
    data = http_json(f"{PLEX_TV}/api/v2/pins?strong=false",
                     headers=_headers(client_id), data={}, method="POST")
    return int(data["id"]), str(data["code"])


def link_url(client_id: str, code: str) -> str:
    q = urllib.parse.urlencode({"clientID": client_id, "code": code,
                                "context[device][product]": _PRODUCT})
    return f"https://app.plex.tv/auth#?{q}"


def poll_pin(client_id: str, pin_id: int) -> str | None:
    """One poll of the PIN; the account token once the user has linked."""
    data = http_json(f"{PLEX_TV}/api/v2/pins/{pin_id}",
                     headers=_headers(client_id))
    token = data.get("authToken")
    return str(token) if token else None


def discover_servers(client_id: str, account_token: str) -> list[dict]:
    """Reachable PMS instances for this account. Each entry:
    ``{name, machine_id, base_url, token}`` — ``token`` is the
    server-specific access token from the resources API, ``base_url`` the
    fastest reachable connection (LAN preferred).

    Probing is two-pass: the first HTTPS contact on a cold resolver (DNS
    for *.plex.direct + TLS handshake) can exceed a short timeout on
    Windows, so a full miss gets one warm retry. If every probe still
    misses, the best local candidate is returned unprobed — a red status
    LED and refresh_auth() beat a dead-end "no reachable server"."""
    resources = http_json(
        f"{PLEX_TV}/api/v2/resources?includeHttps=1&includeRelay=0",
        headers=_headers(client_id, account_token), timeout=10)
    out = []
    for res in resources if isinstance(resources, list) else []:
        if "server" not in str(res.get("provides", "")):
            continue
        token = res.get("accessToken") or account_token
        conns = sorted(res.get("connections") or [],
                       key=lambda c: (not c.get("local"), bool(c.get("relay"))))
        base = None
        for _attempt in range(2):
            base = next((c["uri"] for c in conns
                         if _reachable(c.get("uri", ""), token)), None)
            if base:
                break
        if not base:
            base = next((c["uri"] for c in conns if c.get("local")), None)
        if base:
            out.append({"name": res.get("name", "Plex Server"),
                        "machine_id": res.get("clientIdentifier", ""),
                        "base_url": base.rstrip("/"), "token": token})
    return out


def _reachable(base: str, token: str) -> bool:
    if not base:
        return False
    try:
        req = urllib.request.Request(
            f"{base}/identity",
            headers={"X-Plex-Token": token, "Accept": "application/json"})
        with urllib.request.urlopen(req, timeout=5):
            return True
    except urllib.error.HTTPError:
        return True         # any HTTP answer (401/403/...) IS reachable
    except (OSError, socket.timeout, Exception):
        return False


# ---- the source ----------------------------------------------------------
class PlexSource(MusicSource):
    type = "plex"

    @property
    def base_url(self) -> str:
        return self.account.get("base_url", "").rstrip("/")

    @property
    def client_id(self) -> str:
        return self.account.get("client_id", "")

    def _token(self) -> str:
        return secrets_store.get(f"{self.id}:token") or ""

    def _get(self, path: str, timeout: float = 8.0):
        return http_json(f"{self.base_url}{path}",
                         headers=_headers(self.client_id, self._token()),
                         timeout=timeout)

    def verify(self) -> bool:
        try:
            self._get("/identity", timeout=3)
            return True
        except SourceError:
            return False

    # ---- browse -------------------------------------------------------
    def browse_root(self) -> list[BrowseItem]:
        return self.browse("sections")

    def browse(self, item_id: str) -> list[BrowseItem]:
        kind, _, key = item_id.partition("/")
        if kind == "sections":
            out = []
            for d in self._dir("/library/sections"):
                if d.get("type") == "artist":     # a Plex *music* library
                    out.append(BrowseItem(f"sec/{d['key']}",
                                          d.get("title", "Music"), "folder"))
            out.append(BrowseItem("playlists", "Playlists", "folder"))
            return out
        if kind == "sec":
            return [BrowseItem(f"artist/{d['ratingKey']}",
                               d.get("title", "?"), "artist")
                    for d in self._dir(f"/library/sections/{key}/all?type=8")]
        if kind == "artist":
            return [BrowseItem(f"album/{d['ratingKey']}",
                               d.get("title", "?"), "album")
                    for d in self._dir(f"/library/metadata/{key}/children")]
        if kind == "album":
            return [self._track_item(d)
                    for d in self._dir(f"/library/metadata/{key}/children")
                    if d.get("type") == "track"]
        if kind == "playlists":
            return [BrowseItem(f"playlist/{d['ratingKey']}",
                               d.get("title", "?"), "playlist")
                    for d in self._dir("/playlists?playlistType=audio")]
        if kind == "playlist":
            return [self._track_item(d)
                    for d in self._dir(f"/playlists/{key}/items")
                    if d.get("type") == "track"]
        raise SourceError(f"unknown browse id {item_id!r}")

    def _dir(self, path: str, page: int = 1000) -> list[dict]:
        # paged: a full-library playlist can hold many thousands of items,
        # and one giant response is slow to fetch and parse
        out: list[dict] = []
        start = 0
        while True:
            sep = "&" if "?" in path else "?"
            data = self._get(f"{path}{sep}X-Plex-Container-Start={start}"
                             f"&X-Plex-Container-Size={page}", timeout=20)
            mc = (data.get("MediaContainer") or {})
            # sections arrive under Directory; artists/albums/tracks/
            # playlists under Metadata — merge so every level browses
            chunk = (mc.get("Metadata") or []) + (mc.get("Directory") or [])
            out += chunk
            total = int(mc.get("totalSize", mc.get("size", len(out))) or 0)
            start += len(chunk)
            if not chunk or start >= total:
                return out

    def _track_item(self, d: dict) -> BrowseItem:
        track = self._make_track(d)
        return BrowseItem(f"track/{d['ratingKey']}",
                          track.display(), "track", track=track)

    def _make_track(self, d: dict) -> Track:
        part = ((d.get("Media") or [{}])[0].get("Part") or [{}])[0]
        return Track(
            uri=f"{self.base_url}{part.get('key', '')}",
            title=d.get("title", ""),
            artist=d.get("grandparentTitle", ""),
            album=d.get("parentTitle", ""),
            duration=int(d.get("duration", 0)) // 1000,
            source_id=self.id,
            item_id=str(d.get("ratingKey", "")),
            artwork=d.get("thumb", ""))

    # ---- playback -----------------------------------------------------
    def stream_uri(self, track: Track) -> tuple[str, dict | None]:
        sep = "&" if "?" in track.uri else "?"
        return f"{track.uri}{sep}X-Plex-Token={self._token()}", None

    def artwork_url(self, item: BrowseItem, size: int = 128) -> str:
        thumb = item.track.artwork if item.track else ""
        if not thumb:
            return ""
        q = urllib.parse.urlencode({"width": size, "height": size,
                                    "url": thumb,
                                    "X-Plex-Token": self._token()})
        return f"{self.base_url}/photo/:/transcode?{q}"

    # ---- recovery -----------------------------------------------------
    def refresh_auth(self) -> bool:
        """The usual failure is a changed LAN address, not token expiry:
        re-discover this machine's connections and re-probe."""
        if self.verify():
            return True     # server fine; the track itself may need resolve()
        try:
            for srv in discover_servers(self.client_id, self._token()):
                if srv["machine_id"] == self.account.get("machine_id"):
                    self.account["base_url"] = srv["base_url"]
                    secrets_store.set(f"{self.id}:token", srv["token"])
                    from . import registry
                    registry.save()
                    return True
        except SourceError:
            pass
        return False

    def resolve(self, track: Track) -> Track | None:
        if not track.item_id:
            return track
        try:
            items = self._dir(f"/library/metadata/{track.item_id}")
        except SourceError:
            return None
        return self._make_track(items[0]) if items else None

File: retro/src/plex.c
#include "plex.h"
#include "json.h"
#include "net.h"
#include "app.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef _WIN32
#define snprintf _snprintf
#endif

static void headers(char *out, int cap, const char *client_id, const char *token)
{
    int n = snprintf(out, (size_t)cap,
                     "X-Plex-Product: EasyAmp\r\nX-Plex-Version: " EA_VERSION "\r\nX-Plex-Client-Identifier: %s\r\n"
                     "X-Plex-Device-Name: EasyAmp retro\r\nX-Plex-Platform: Windows\r\n", client_id);
    if (token && token[0] && n > 0 && n < cap) snprintf(out + n, (size_t)(cap - n), "X-Plex-Token: %s\r\n", token);
    out[cap - 1] = 0;
}

static void fail(char *err, int cap, const char *what, const ea_http *r)
{
    if (r->status) snprintf(err, (size_t)cap, "%s: HTTP %d", what, r->status);
    else snprintf(err, (size_t)cap, "%s: %s", what, r->err);
    err[cap - 1] = 0;
}

int plex_pin_start(const char *client_id, long *pin_id, char code[8], char *err, int errcap)
{
    char h[512];
    ea_http r;
    ea_json j;
    int ok = 0;
    headers(h, sizeof h, client_id, 0);
    if (net_request("POST", "https://plex.tv/api/v2/pins?strong=false", h, "", &r, 20000) != 201) { fail(err, errcap, "plex.tv", &r); net_free(&r); return 0; }
    if (json_parse(&j, r.body, r.len)) {
        int root = json_root(&j), c = json_get(&j, root, "code");
        *pin_id = json_int(&j, json_get(&j, root, "id"));
        if (json_is_string(&j, c) && *pin_id) { json_str(&j, c, code, 8); ok = 1; }
        json_free(&j);
    }
    if (!ok) snprintf(err, (size_t)errcap, "plex.tv sent an unexpected reply");
    net_free(&r);
    return ok;
}

int plex_pin_poll(const char *client_id, long pin_id, char *token, int tokcap, char *err, int errcap)
{
    char h[512], url[128];
    ea_http r;
    ea_json j;
    int res = 0, st;
    headers(h, sizeof h, client_id, 0);
    sprintf(url, "https://plex.tv/api/v2/pins/%ld", pin_id);
    st = net_request("GET", url, h, 0, &r, 20000);
    if (st == 404) { snprintf(err, (size_t)errcap, "code expired - try again"); net_free(&r); return -1; }
    if (st != 200) { fail(err, errcap, "plex.tv", &r); net_free(&r); return st ? -1 : 0; }   /* a dropped poll is not fatal */
    if (json_parse(&j, r.body, r.len)) {
        int t = json_get(&j, json_root(&j), "authToken");
        if (json_is_string(&j, t)) { json_str(&j, t, token, tokcap); res = token[0] ? 1 : 0; }
        json_free(&j);
    }
    net_free(&r);
    return res;
}

int plex_alive(const plex_server *s)
{
    char url[160], h[128];
    ea_http r;
    int st;
    snprintf(url, sizeof url, "%s/identity", s->base);
    snprintf(h, sizeof h, "X-Plex-Token: %s\r\n", s->token);
    st = net_request("GET", url, h, 0, &r, 4000);
    net_free(&r);
    return st != 0;                  /* any HTTP answer, even 401, proves it is reachable */
}

int plex_discover(const char *client_id, const char *account_token, plex_server *out, int max, char *err, int errcap)
{
    char h[640];
    ea_http r;
    ea_json j;
    int found = 0, saw_server = 0, res;
    headers(h, sizeof h, client_id, account_token);
    if (net_request("GET", "https://plex.tv/api/v2/resources?includeHttps=1&includeRelay=0", h, 0, &r, 25000) != 200) {
        fail(err, errcap, "plex.tv", &r); net_free(&r); return 0;
    }
    if (!json_parse(&j, r.body, r.len)) { snprintf(err, (size_t)errcap, "plex.tv sent an unexpected reply"); net_free(&r); return 0; }
    for (res = json_first(&j, json_root(&j)); res >= 0 && found < max; res = json_next(&j, res)) {
        char provides[64], addr[64], tmp[16];
        int conns, c, pass;
        plex_server s;
        json_str(&j, json_get(&j, res, "provides"), provides, sizeof provides);
        if (!strstr(provides, "server")) continue;
        saw_server = 1;
        memset(&s, 0, sizeof s);
        json_str(&j, json_get(&j, res, "name"), s.name, sizeof s.name);
        json_str(&j, json_get(&j, res, "clientIdentifier"), s.machine, sizeof s.machine);
        if (json_is_string(&j, json_get(&j, res, "accessToken"))) json_str(&j, json_get(&j, res, "accessToken"), s.token, sizeof s.token);
        else strncpy(s.token, account_token, sizeof s.token - 1);
        conns = json_get(&j, res, "connections");
        /* LAN addresses first, over plain HTTP: no TLS cost per request, and
         * the playback stream can use the very same base URL */
        for (pass = 0; pass < 2 && !s.base[0]; pass++)
            for (c = json_first(&j, conns); c >= 0; c = json_next(&j, c)) {
                int local = json_true(&j, json_get(&j, c, "local")), relay = json_true(&j, json_get(&j, c, "relay"));
                if (relay || local != (pass == 0)) continue;
                json_str(&j, json_get(&j, c, "address"), addr, sizeof addr);
                json_str(&j, json_get(&j, c, "port"), tmp, sizeof tmp);
                if (!addr[0] || strchr(addr, ':')) continue;              /* no IPv6 on Windows 98 */
                snprintf(s.base, sizeof s.base, "http://%s:%s", addr, tmp[0] ? tmp : "32400");
                if (plex_alive(&s)) break;
                s.base[0] = 0;
            }
        if (s.base[0]) out[found++] = s;
    }
    json_free(&j);
    net_free(&r);
    if (!found) snprintf(err, (size_t)errcap, saw_server ? "server found but not reachable over plain HTTP (Settings > Network > Secure connections: Preferred)"
                                                         : "no Plex Media Server on this account");
    return found;
}

static void track_from(const ea_json *j, int d, plex_item *it)
{
    int media = json_first(j, json_get(j, d, "Media")), part = json_first(j, json_get(j, media, "Part"));
    char title[96];
    it->kind = PLEX_TRACK;
    json_str(j, json_get(j, d, "ratingKey"), it->key, sizeof it->key);
    json_str(j, json_get(j, d, "title"), title, sizeof title);
    json_str(j, json_get(j, d, "grandparentTitle"), it->artist, sizeof it->artist);
    if (json_get(j, d, "originalTitle") >= 0) json_str(j, json_get(j, d, "originalTitle"), it->artist, sizeof it->artist);
    snprintf(it->name, sizeof it->name, "%s%s%s", it->artist, it->artist[0] ? " - " : "", title);
    it->name[sizeof it->name - 1] = 0;
    json_str(j, json_get(j, part, "key"), it->part, sizeof it->part);
    json_str(j, json_get(j, media, "audioCodec"), it->codec, sizeof it->codec);
    it->dur_s = (int)(json_int(j, json_get(j, d, "duration")) / 1000);
}

plex_item *plex_browse(const plex_server *s, const char *client_id, const char *node, int *n, char *err, int errcap)
{
    char url[256], h[640], list[16] = "Metadata";
    ea_http r;
    ea_json j;
    plex_item *items = 0;
    int kind, mc, arr, d, count, i = 0;
    *n = 0;
    if (!node[0]) { snprintf(url, sizeof url, "%s/library/sections", s->base); kind = PLEX_SECTION; strcpy(list, "Directory"); }
    else if (!strncmp(node, "section/", 8)) { snprintf(url, sizeof url, "%s/library/sections/%s/all?type=8", s->base, node + 8); kind = PLEX_ARTIST; }
    else if (!strncmp(node, "artist/", 7)) { snprintf(url, sizeof url, "%s/library/metadata/%s/children", s->base, node + 7); kind = PLEX_ALBUM; }
    else if (!strncmp(node, "album/", 6)) { snprintf(url, sizeof url, "%s/library/metadata/%s/children", s->base, node + 6); kind = PLEX_TRACK; }
    else { snprintf(err, (size_t)errcap, "bad node"); return 0; }
    headers(h, sizeof h, client_id, s->token);
    if (net_request("GET", url, h, 0, &r, 30000) != 200) { fail(err, errcap, s->name, &r); net_free(&r); return 0; }
    if (!json_parse(&j, r.body, r.len)) { snprintf(err, (size_t)errcap, "server sent an unexpected reply"); net_free(&r); return 0; }
    mc = json_get(&j, json_root(&j), "MediaContainer");
    arr = json_get(&j, mc, list);
    count = json_size(&j, arr);
    items = (plex_item *)calloc((size_t)(count > 0 ? count : 1), sizeof *items);
    for (d = json_first(&j, arr); items && d >= 0 && i < count; d = json_next(&j, d)) {
        plex_item *it = &items[i];
        if (kind == PLEX_SECTION) {
            char type[16];
            json_str(&j, json_get(&j, d, "type"), type, sizeof type);
            if (strcmp(type, "artist")) continue;                       /* music libraries only */
            json_str(&j, json_get(&j, d, "key"), it->key, sizeof it->key);
            json_str(&j, json_get(&j, d, "title"), it->name, sizeof it->name);
            it->kind = PLEX_SECTION;
        } else if (kind == PLEX_TRACK) track_from(&j, d, it);
        else {
            it->kind = kind;
            json_str(&j, json_get(&j, d, "ratingKey"), it->key, sizeof it->key);
            json_str(&j, json_get(&j, d, "title"), it->name, sizeof it->name);
        }
        i++;
    }
    json_free(&j);
    net_free(&r);
    *n = i;
    if (!i && items) err[0] = 0;
    return items;
}

static int is_mp3(const char *codec)
{
    return !codec[0] || ((codec[0] | 32) == 'm' && (codec[1] | 32) == 'p' && codec[2] == '3' && !codec[3]);
}

void plex_track_url(const plex_server *s, const char *client_id, const plex_item *t, char *out, int cap)
{
    if (is_mp3(t->codec)) snprintf(out, (size_t)cap, "%s%s", s->base, t->part);
    else snprintf(out, (size_t)cap,
                  "%s/music/:/transcode/universal/start.mp3?path=%%2Flibrary%%2Fmetadata%%2F%s&mediaIndex=0&partIndex=0"
                  "&protocol=http&directPlay=0&directStream=0&audioCodec=mp3&maxAudioBitrate=192&X-Plex-Platform=Chrome"
                  "&X-Plex-Client-Identifier=%s&X-Plex-Session-Identifier=%s-%s"
                  "&X-Plex-Client-Profile-Extra=add-transcode-target%%28type%%3DmusicProfile%%26context%%3Dstreaming%%26protocol%%3Dhttp%%26container%%3Dmp3%%26audioCodec%%3Dmp3%%29",
                  s->base, t->key, client_id, client_id, t->key);
    out[cap - 1] = 0;
}

void plex_auth_url(const plex_server *s, const char *url, char *out, int cap)
{
    snprintf(out, (size_t)cap, "%s%sX-Plex-Token=%s", url, strchr(url, '?') ? "&" : "?", s->token);
    out[cap - 1] = 0;
}
